platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r89/ToolingApiPolymorphismCrossVersionTest.groovy

```diff
@@ -0,0 +1,207 @@
+                        implementationClass = "my.MyPlugin"
+                    }
+                }
+            }
+        """
+
+        file("plugins/src/main/groovy/my/MyModel.groovy") << """
+            package org.gradle.integtests.tooling.fixture
+
+            interface BaseModel {
+            }
+
+            interface ShallowChildModel extends BaseModel {
+                String getShallowMessage()
+            }
+
+            interface DeepChildModel extends BaseModel {
+                String getDeepMessage()
+            }
+
+            interface SideModel extends BaseModel {
+            }
+
+            abstract class AbstractModel implements DeepChildModel {
+            }
+
+            class DefaultModel extends AbstractModel implements ShallowChildModel, SideModel, java.io.Serializable {
+                private final String message
+                DefaultModel(String message) { this.message = message }
+                String getShallowMessage() { "shallow " + message }
+                String getDeepMessage() { "deep " + message }
+                String toString() { message }
+            }
+
+            class DefaultCompositeModel implements java.io.Serializable {
+                private final DefaultModel nested
+                DefaultCompositeModel(DefaultModel nested) { this.nested = nested }
+                BaseModel getNested() { nested }
+            }
+        """
+    }
+
+    def "supports model polymorphism"() {
+        given:
+        file("plugins/src/main/groovy/my/MyModelBuilder.groovy") << """
+            package my
+
+            import ${ToolingModelBuilder.name}
+            import ${Project.name}
+
+            class MyModelBuilder implements ToolingModelBuilder {
+                boolean canBuild(String modelName) {
+                    return modelName == "${BaseModel.name}"
+                }
+                Object buildAll(String modelName, Project project) {
+                    println("creating model for \$project")
+                    return new org.gradle.integtests.tooling.fixture.DefaultModel("poly from '" + project.name + "'")
+                }
+            }
+        """.stripIndent()
+
+        addBuilderRegisteringPluginImplementation("plugins", "MyModelBuilder")
+
+        settingsFile << """
+            includeBuild("plugins")
+            rootProject.name = 'root'
+        """
+        buildFile << """
+            plugins {
+                id("my.plugin")
+            }
+        """
+
+        when:
+        def model = toolingApi.withConnection() { connection -> connection.getModel(BaseModel) }
+
+        then:
+        model != null
+
+        model instanceof BaseModel
+
+        model instanceof ShallowChildModel
+        ((ShallowChildModel) model).getShallowMessage() == "shallow poly from 'root'"
+
+        model instanceof DeepChildModel
+        ((DeepChildModel) model).getDeepMessage() == "deep poly from 'root'"
+
+        !(model instanceof SideModel)
```

Comment:
💭 Would it be useful to extract these checks from this test and the one below?

Reply:
Done.